src/esp/physics/PhysicsManager.h

```diff
@@ -0,0 +1,210 @@
+  void setFrictionCoefficient(const int physObjectID,
+                              const double frictionCoefficient);
+  void setRestitutionCoefficient(const int physObjectID,
+                                 const double restitutionCoefficient);
+  void setLinearDamping(const int physObjectID, const double linDamping);
+  void setAngularDamping(const int physObjectID, const double angDamping);
+
+  // ============ Object Getter functions =============
+  const double getMass(const int physObjectID);
+  const Magnum::Vector3 getCOM(const int physObjectID);
+  const Magnum::Vector3 getInertia(const int physObjectID);
+  const double getScale(const int physObjectID);
+  const double getFrictionCoefficient(const int physObjectID);
+  const double getRestitutionCoefficient(const int physObjectID);
+  const double getLinearDamping(const int physObjectID);
+  const double getAngularDamping(const int physObjectID);
+
+  // ============= Platform dependent function =============
+  virtual const double getMargin(const int physObjectID) { return 0.0; };
+  virtual void setMargin(const int physObjectID, const double margin){};
+
+  // =========== Debug functions ===========
+  int checkActiveObjects();
+
+  //============ Interact with objects =============
+  // Alex NOTE: engine specifics handled by objects themselves...
+  void applyForce(const int physObjectID,
+                  Magnum::Vector3& force,
+                  Magnum::Vector3& relPos);
+
+  void applyImpulse(const int physObjectID,
+                    Magnum::Vector3& impulse,
+                    Magnum::Vector3& relPos);
+  /*
+  Magnum::SceneGraph::DrawableGroup3D& getDrawables() { return debugDrawables; }
+  const Magnum::SceneGraph::DrawableGroup3D& getDrawables() const {
+    return debugDrawables;
+  }
+  */
+
+ protected:
+  //! Check if mesh primitive type is valid for bullet physics engine
+  virtual bool isMeshPrimitiveValid(const assets::CollisionMeshData& meshData);
+
+  // acquires an ID from either the recycledObjectIDs_ or by incrementing
+  // nextObjectID_
+  int allocateObjectID();
+
+  // recycle an physObjectID
+  int deallocateObjectID(int physObjectID);
+
+  //! Create and initialize rigid object
+  virtual int makeRigidObject(
+      const std::vector<assets::CollisionMeshData>& meshGroup,
+      assets::PhysicsObjectAttributes physicsObjectAttributes);
+
+  // use this to instantiate physics objects from the physicsObjectLibrary_
+  assets::ResourceManager* resourceManager_;
+
+  //! ==== physics engines ====
+  enum PhysicsSimulationLibrary { NONE, BULLET };
+  PhysicsSimulationLibrary activePhysSimLib_ = NONE;  // default
+
+  //! Used to keep track of all sceneNodes that have physical properties
+  scene::SceneNode* physicsNode_ = nullptr;
+  std::shared_ptr<physics::RigidObject> sceneNode_ = nullptr;
+
+  //! ==== dynamic object resources ===
+  std::map<int, std::shared_ptr<physics::RigidObject>> existingObjects_;
```

Comment:
`RigidObject` current inherits from a `Magnum::SceneGraph::Object` and thus can't be shared nor can you own that memory, so anywhere with `std::shared_ptr<RigidObject>` will cause a double free.  @mosra I believe making `RigidObject` inherit from `Magnum::SceneGraph::AbstractFeature3D` is the correct way to allow for a persistent reference to a `Magnum::SceneGraph::Object`?

Reply:
This will end up being fine as long as the parent object doesn't get deleted before any of the objects in the list.  A map to `unique_ptr` _might_ end up being okay, it isn't really a unique ptr, but it might be okay.

Reply:
Idea:  We store the `physics::RigidObject` just as raw pointers.  Then we have a physics manager per scene and the resource manager is then responsible for tying the lifetime of the physics manager and the scene graph it is managing objects within.

Reply:
> I believe making `RigidObject` inherit from `Magnum::SceneGraph::AbstractFeature3D` is the correct way to allow for a persistent reference to a `Magnum::SceneGraph::Object`?

Yes, that's the designed way to do that. I would then rename it to not have `Object` in its name, tho, to avoid confusing inheritance assumptions. (`RigidBody` for example?)

> fine as long as the parent object doesn't get deleted before any of the objects in the list

While working on the Python bindings I made it possible to override this implicit memory management behavior if you implement your own `doErase()` on either the object or the feature ([docs](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1LinkedListItem.html#acb954b145234f6ff5ae43f9f2a951b0e), hopefully understandable enough). In this case, if it gets implemented on the feature, the feature stays alive even after its owner object gets destroyed and then the proposed `unique_ptr` would take care of its deletion when appropriate.